gitindex/index.go

```diff
@@ -532,6 +528,248 @@ func newIgnoreMatcher(tree *object.Tree) (*ignore.Matcher, error) {
+		if err != nil {
+			return nil, fmt.Errorf("getting lasted indexed git tree for branch %q: %w", branch.Name, err)
+		}
+
+		changes, err := lastIndexedTree.Diff(branchToCurrentTree[branch.Name])
+		if err != nil {
+			return nil, fmt.Errorf("generating changeset for branch %q: %w", branch.Name, err)
+		}
+
+		patch, err := changes.Patch()
+		if err != nil {
+			return nil, fmt.Errorf("generating patch from changeset for branch %q: %w", branch.Name, err)
+		}
+
+		for _, fp := range patch.FilePatches() {
+			fromFile, toFile := fp.Files()
+
+			if fromFile == nil {
+				// file added
+				file := fileKey{Path: toFile.Path(), ID: toFile.Hash()}
+				repos[file] = hackSharedBlobLocation
+				branchMap[file] = append(branchMap[file], branch.Name)
+				continue
+			}
+
+			changedOrDeletedFile := fileKey{Path: fromFile.Path(), ID: fromFile.Hash()}
+
+			// If the file is either modified or deleted, we need to add ALL versions
+			// of this file (across all branches) to the build
+			for b, currentTree := range branchToCurrentTree {
+				f, err := currentTree.File(changedOrDeletedFile.Path)
+				if err != nil {
+					// the file doesn't exist in this branch
+					if errors.Is(err, object.ErrFileNotFound) {
+						continue
+					}
+
+					return nil, fmt.Errorf("getting hash for file %q in branch %q: %w", changedOrDeletedFile.Path, b, err)
+				}
+
+				file := fileKey{Path: changedOrDeletedFile.Path, ID: f.ID()}
+
+				repos[file] = hackSharedBlobLocation
+				branchMap[file] = append(branchMap[file], b)
+			}
+
+			changedOrDeletedPaths = append(changedOrDeletedPaths, changedOrDeletedFile.Path)
+		}
+	}
+
+	// we need to de-duplicate the branch map before returning it - it's possible for the same
+	// branch to have been added multiple times if a file has been modified across multiple commits
+
+	for file, branches := range branchMap {
+		sort.Strings(branches)
+		branchMap[file] = uniq(branches)
+	}
+
+	// we also need to de-duplicate the list of changed or deleted file paths, it's also possible to have duplicates
+	// for the same reasoning as above
+
+	sort.Strings(changedOrDeletedPaths)
+	changedOrDeletedPaths = uniq(changedOrDeletedPaths)
+
+	return &buildOutput{
+		repos:                 repos,
+		branchMap:             branchMap,
+		branchVersions:        nil,
+		changedOrDeletedPaths: changedOrDeletedPaths,
+	}, nil
+}
+
+func prepareNormalBuild(options *Options, repository *git.Repository) (*buildOutput, error) {
+	var repoCache *RepoCache
+	if options.Submodules {
+		repoCache = NewRepoCache(options.RepoCacheDir)
+	}
+
+	// branch => (path, sha1) => repo.
+	repos := map[fileKey]BlobLocation{}
+
+	// fileKey => branches
+	branchMap := map[fileKey][]string{}
+
+	// Branch => Repo => SHA1
+	branchVersions := map[string]map[string]plumbing.Hash{}
+
+	branches, err := expandBranches(repository, options.Branches, options.BranchPrefix)
+	if err != nil {
+		return nil, fmt.Errorf("expandBranches: %w", err)
+	}
+
+	for _, b := range branches {
```

Comment:
Now we loop over branches twice. Is there a way we could avoid this? In practice this shouldn't have any noticable impact I guess.